cli/internal/fs/turbo_json.go

```diff
@@ -162,18 +163,34 @@ func (c *TaskDefinition) UnmarshalJSON(data []byte) error {
 	} else {
 		c.ShouldCache = *rawPipeline.Cache
 	}
-	c.EnvVarDependencies = []string{}
+
+	envVarDependencies := make(util.Set)
 	c.TopologicalDependencies = []string{}
 	c.TaskDependencies = []string{}
+
 	for _, dependency := range rawPipeline.DependsOn {
 		if strings.HasPrefix(dependency, envPipelineDelimiter) {
-			c.EnvVarDependencies = append(c.EnvVarDependencies, strings.TrimPrefix(dependency, envPipelineDelimiter))
+			// TODO: It would be nice to include the taskID in this log message, but we don't have access
+			// to that with the raw pipelineJSON struct.
+			log.Printf("[DEPRECATED] Declaring environment variables in \"dependsOn\" is deprecated, found %s. Use the \"env\" key instead.\n", dependency)
+			envVarDependencies.Add(strings.TrimPrefix(dependency, envPipelineDelimiter))
 		} else if strings.HasPrefix(dependency, topologicalPipelineDelimiter) {
 			c.TopologicalDependencies = append(c.TopologicalDependencies, strings.TrimPrefix(dependency, topologicalPipelineDelimiter))
 		} else {
 			c.TaskDependencies = append(c.TaskDependencies, dependency)
 		}
 	}
+
+	// Append env key into EnvVarDependencies
+	for _, value := range rawPipeline.Env {
+		if strings.HasPrefix(value, envPipelineDelimiter) {
+			log.Printf("[WARNING] Ignoring \"%s\" environment variable declaration. Remove %s prefix.", value, envPipelineDelimiter)
+		} else {
+			envVarDependencies.Add(value)
+		}
+	}
```

Comment:
I slightly prefer strict behavior but since this has its own pool I'm not sure that this gives the end-user any value and just unnecessarily gets in their way.

Reply:
Are you saying you *want* to support with and without `$` prefix? What's the strict behavior you prefer?

Reply:
Strict behavior would be: leave as is. This is the alternative:

```suggestion
	envVarDependencies.Add(strings.TrimPrefix(value), envPipelineDelimiter)
```

Reply:
/cc @vercel/turbo-oss some prime bikeshedding here.

Reply:
Got it. @gsoltis and I talked about this while designing the expanded version of the config changes and he felt that `$FOO` in a key called `env` was redundant. Personally, I think I _like_ seeing `$FOO` instead of `FOO` for a couple reasons

- it immediately tells me it's an environment variable
- semantically, we are _reading_ the variable, rather than writing, so similar to `echo $FOO`, it makes more sense to refer to it with the dollar sign.

I don't have a strong opinion though, and don't think my reasoning is foolproof, so I would be ok ending up picking either one. I **don't** think we should allow both though.

Reply:
Part of the motivation for making this change is to get away from in-band signaling / micro-syntaxes. As such, I think it's best if the entries in `env` are treated without introspection. If you have a variable named `FOO`, put `FOO`. If you put `$FOO`, we will look for an environment variable named `$FOO`.

It makes for a simple explanation of what to put in that field, and no magic.

Reply:
I was going to do a hard error, but because this is per-task running concurrently, I didn't want to spend time figuring out:

- how this would behave for the user or how we're handling that upstream of this
- how to write a test that expects a fatal error

Both surmountable problems to be sure if that's better behavior!

Reply:
So, @gsoltis's version from the comment:

```suggestion
	envVarDependencies.Add(value, envPipelineDelimiter)
```

Reply:
Or a variation combining @gsoltis and @mehulkar's versions:
```suggestion
	for _, value := range rawPipeline.Env {
		if strings.HasPrefix(value, envPipelineDelimiter) {
			log.Printf("[WARNING] You specified $ as an environment variable prefix. We're treating \"%s\" as a literal environment variable name.", value)
		}
		envVarDependencies.Add(value)
	}
```

Reply:
In order from me:
1. [Automatically remove $ prefix](https://github.com/vercel/turborepo/pull/1932#discussion_r969929521)
2. [Literal to input + warning for $ prefix](https://github.com/vercel/turborepo/pull/1932#discussion_r970462879)
3. ...
4. ...
5. ...
100. [Literal to input, no warning](https://github.com/vercel/turborepo/pull/1932#discussion_r970460564)

I don't consider the currently implemented approach (reject $-prefixed + warning) to be a viable solution.

Reply:
I think #2, is a good suggestion. Treat it as a literal, but add a warning. I'll add a code note as well explaining how we used to want users to declare `$FOO`, but then we changed it so now we try to be helpful.